vault/logical_system.go

```diff
@@ -5542,6 +5545,397 @@ func checkListingVisibility(visibility ListingVisibilityType) error {
 	return nil
 }
 
+func (b *SystemBackend) handleListCustomMessages(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	var filters ListUICustomMessagesFilters
+
+	authenticatedValue, ok, err := d.GetOkErr("authenticated")
+	if err != nil {
+		return logical.ErrorResponse("invalid authenticated parameter value: %s", err), nil
+	}
+
+	if ok {
+		filters.Authenticated(authenticatedValue.(bool))
+	}
+
+	typeValue, ok, err := d.GetOkErr("type")
+	if err != nil {
+		return logical.ErrorResponse("invalid type parameter value: %s", err), nil
+	}
+
+	if ok {
+		filters.MessageType(typeValue.(string))
+	}
+
+	activeValue, ok, err := d.GetOkErr("active")
+	if err != nil {
+		return logical.ErrorResponse("invalid active parameter value: %s", err), nil
+	}
+
+	if ok {
+		filters.Active(activeValue.(bool))
+	}
+
+	entries, err := b.Core.uiConfig.ListCustomMessages(ctx, filters)
+	if err != nil {
+		return logical.ErrorResponse("failed to retrieve list of custom messages: %s", err), nil
+	}
+
+	messages := make(map[string]any)
+	keys := make([]string, len(entries))
+
+	for i, entry := range entries {
+		entryMap := make(map[string]any)
+		entryMap["title"] = entry.Title
+		entryMap["type"] = entry.MessageType
+		entryMap["authenticated"] = entry.Authenticated
+		entryMap["start_time"] = entry.StartTime
+		entryMap["end_time"] = entry.EndTime
+		entryMap["active"] = entry.active
+
+		messages[entry.Id] = entryMap
+		keys[i] = entry.Id
+	}
+
+	return logical.ListResponseWithInfo(keys, messages), nil
+}
+
+func (b *SystemBackend) handleCreateCustomMessages(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	var err error
+
+	var title string
+	titleValue, ok, err := d.GetOkErr("title")
+	if err != nil {
+		return logical.ErrorResponse("invalid title parameter value: %s", err), nil
```

Comment:
This is a nit, totally optional, but would `%q` be better than `%s` here?

Reply:
Do we typically quote the error messages?

Reply:
Oh, woops, you're right. I totally misread this as us giving them back the title in this situation. Feel free to ignore this comment, resolving it!